crates/devices/virtio-console/src/console.rs

```diff
@@ -0,0 +1,516 @@
+
+        match self {
+            WriteToGuestFailed(ref err) => {
+                write!(f, "Console failed to write slice to guest memory: {}", err)
+            }
+            WriteToOutputFailed(ref err) => {
+                write!(f, "Console failed to write data to output sink: {}", err)
+            }
+            ChainLengthOverflow => {
+                write!(f, "Console transmitq chain length has overflown usize.")
+            }
+            BufferCapacityExceeded => {
+                write!(
+                    f,
+                    "Console input buffer maximum capacity has been exceeded."
+                )
+            }
+            UnexpectedReadOnlyDescriptor => {
+                write!(f, "Unexpected read-only descriptor.")
+            }
+            UnexpectedWriteOnlyDescriptor => {
+                write!(f, "Unexpected write-only descriptor.")
+            }
+            InvalidBufferCapacity => {
+                write!(f, "The maximum capacity of the input buffer cannot be 0.")
+            }
+            OutputSinkFlushFailed(ref err) => {
+                write!(f, "Output sink flush has not written all bytes: {}", err)
+            }
+        }
+    }
+}
+
+/// Console struct that implements the abstraction of virtio console descriptor chain handling.
+#[derive(Clone, Debug)]
+pub struct Console<T: Write> {
+    /// Buffer that stores data to be sent to the driver.
+    input_buffer: Arc<Mutex<Vec<u8>>>,
+    /// Capacity of the input buffer.
+    capacity: usize,
+    /// Output sink.
+    output: T,
+}
+
+// Default maximum capacity
+// TODO: Find the optimal CAPACITY and MAX_CAPACITY through testing
+const MAX_CAPACITY: usize = 16384;
+const DEFAULT_CAPACITY: usize = 4096;
+
+impl Default for Console<Stdout> {
+    fn default() -> Self {
+        Self::new_with_capacity(DEFAULT_CAPACITY, stdout()).unwrap()
+    }
+}
+
+/// Used to check if two consoles are of the same size.
+impl<T> PartialEq for Console<T>
+where
+    T: Write,
+{
+    fn eq(&self, other: &Self) -> bool {
+        self.capacity == other.capacity
+    }
+}
+
+impl<T> Console<T>
+where
+    T: Write,
+{
+    /// Create new console object with the default `capacity`.
+    ///
+    /// # Arguments
+    /// * `output` - The output sink where all the data received from
+    ///              the transmitq will be written.
+    /// # Returns
+    /// The console object.
+    pub fn new(output: T) -> Self {
+        Self::new_with_capacity(DEFAULT_CAPACITY, output).unwrap()
+    }
+
+    /// Create new console object with specified `capacity`.
+    ///
+    /// # Arguments
+    /// * `capacity` - The maximum capacity of the input buffer.
+    ///                    Passing 0 will return an `Error`.
+    /// * `output` - The output sink where all the data received from
+    ///              the transmitq will be written.
+    /// # Returns
+    /// The console object or an `Error` if the capacity is 0 or higher than `MAX_CAPACITY`.
```

Comment:
The users of the interface cannot know what `MAX_CAPACITY` is unless it is exported. We should mark it as public, and we can then use the following so that the value can easily be found while reading the documentation:
```rust
The console object or an `Error` if the capacity is 0 or higher than [`MAX_CAPACITY`].
```